certbot/ocsp.py

```diff
@@ -62,37 +84,104 @@ def ocsp_revoked(self, cert_path, chain_path):
+        request = builder.build()
+        request_binary = request.public_bytes(serialization.Encoding.DER)
+        response = requests.post(url, data=request_binary,
+                                 headers={'Content-Type': 'application/ocsp-request'})
+        if not response.status_code == 200:
+            logger.info("OCSP check failed for %s (are we offline?)", cert_path)
+            return False
+        response_ocsp = ocsp.load_der_ocsp_response(response.content)
 
-    def determine_ocsp_server(self, cert_path):
+        # Check OCSP signature
+        try:
+            _check_ocsp_response_signature(response_ocsp, issuer)
+        except UnsupportedAlgorithm as e:
+            logger.error(str(e))
+            return False
+        except InvalidSignature:
+            logger.error('Invalid signature for OCSP response on %s', cert_path)
+            return False
+
+        # Check OCSP certificate status
+        try:
+            logger.debug("OCSP certificate status for %s is: %s",
+                         cert_path, response_ocsp.certificate_status)
+            return response_ocsp.certificate_status == ocsp.OCSPCertStatus.REVOKED
+        except ValueError:  # pragma: no cover
+            logger.info("Invalid OCSP response status for %s: %s",
+                        cert_path, response_ocsp.response_status)
+            return False
+
+    @staticmethod
+    def _determine_ocsp_server(cert_path):
+        # type: (str) -> Tuple[Optional[str], Optional[str]]
         """Extract the OCSP server host from a certificate.
 
         :param str cert_path: Path to the cert we're checking OCSP for
         :rtype tuple:
         :returns: (OCSP server URL or None, OCSP server host or None)
 
         """
+        with open(cert_path, 'rb') as file_handler:
+            cert = x509.load_pem_x509_certificate(file_handler.read(), default_backend())
         try:
-            url, _err = util.run_script(
-                ["openssl", "x509", "-in", cert_path, "-noout", "-ocsp_uri"],
-                log=logger.debug)
-        except errors.SubprocessError:
+            extension = cert.extensions.get_extension_for_class(x509.AuthorityInformationAccess)
+            ocsp_oid = x509.AuthorityInformationAccessOID.OCSP
+            descriptions = [description for description in extension.value
+                            if description.access_method == ocsp_oid]
+
+            url = descriptions[0].access_location.value
+        except (x509.ExtensionNotFound, IndexError):
             logger.info("Cannot extract OCSP URI from %s", cert_path)
             return None, None
 
         url = url.rstrip()
         host = url.partition("://")[2].rstrip("/")
+
         if host:
             return url, host
         else:
             logger.info("Cannot process OCSP host from URL (%s) in cert at %s", url, cert_path)
             return None, None
 
+
+def _check_ocsp_response_signature(response_ocsp, issuer_cert):
```

Comment:
@reaperhulk, is there a better way for us to write this function? In particular, relying on the internal `_SIG_OIDS_TO_HASH` makes me nervous.

Reply:
It saves a lot of boilerplate code, and seems strongly anchored in `cryptography` codebase. Anyway, waiting for opinions on that matter.

Reply:
This isn't wrong, but it's actually a missing API for `cryptography`. On Certificates, CRLs, and CSRs we have `signature_hash_algorithm`, which returns the value you want here. We should add that to cryptography.

Reply:
Method `signature_hash_algorithm` has been added on upstream, will be available with Cryptography 2.5. I added a TODO in the code to remember that.